### PR TITLE
AGS3 refactor ManagedObjectPool to address slow lookups

### PR DESCRIPTION
**Pull request by sonneveld**
Uses maps for object handle and address lookups.

### CODE REVIEW

File: Engine/ac/dynobj/managedobjectpool.cpp [hunk 1]
```diff
@@ -65,166 +47,155 @@ int32_t ManagedObjectPool::AddRef(int32_t handle) {
...
+
+    objects.insert({handle, {
+        ManagedObject {
+            obj_type: plugin_object ? kScValPluginObject : kScValDynamicObject,
+            handle: handle,
+            addr: address,
+            callback: callback,
+            refCount: 0
+        }
+    }});
```
Comment (ivan-mogilko) [inline Engine/ac/dynobj/managedobjectpool.cpp]: This does not compile using MSVS 2015 compiler. What version of standard is this kind of initialization?

Comment (sonneveld) [reply Engine/ac/dynobj/managedobjectpool.cpp]: Interesting. I'm using MSVS 2017.  It's possible I didn't test this under Windows and I've inadvertently used a gcc extension. I was aiming for this: https://docs.microsoft.com/en-us/cpp/cpp/initializing-classes-and-structs-without-constructors-cpp?view=vs-2017

Comment (sonneveld) [reply Engine/ac/dynobj/managedobjectpool.cpp]: Pushed a fix.

**Comment by sonneveld** [comment]
Rebased and fixed conflicts.

**Comment by ivan-mogilko** [comment]
Although I was planning similar change myself (use hashmap for address -> handle lookup), but after testing the results are not simple. Profiling shows that in regular games AddressToHandle time decreased.
But I also tested this with an experimental game which does a lot of vector/matrix calculations in script, passing them as dynamic arrays with a very short lifespan. This test displays certain downgrade in perfomance. Using the "--fps" arg to max out frames, the game does up to 74 fps with AGS 3.5.0.7 release and only around 65-67 fps with this branch.
It's hard to rely on fps test of course but I have a guess this may be related to insertion and deletion of hashmap elements. In original code elements of array were not deallocated, only reallocated if array is too small.

**Comment by ivan-mogilko** [comment]
I tried increasing the GARBAGE_COLLECTION_INTERVAL constant but it did not do much effect.

But I've got an impression that GC is not working correctly at the moment. The element removal is done at SubRef call. Maybe it should be done differently.

**Comment by sonneveld** [comment]
I've changed how the objects are stored. It's closer to the original with an index to a vector now. Please let me know how it goes with your test game when you have time.

**Comment by sonneveld** [comment]
Just pushed a small fix.

**Pull request merged**